cmd/zrepl/mgmt_conn.c

```diff
@@ -0,0 +1,899 @@
+
+	mutex_enter(&conn_list_mtx);
+	SLIST_FOREACH(conn, &uzfs_mgmt_conns, conn_next) {
+		/* we need to create new connection */
+		if (conn->conn_refcount > 0 && conn->conn_fd < 0 &&
+		    time(NULL) - conn->conn_last_connect > RECONNECT_DELAY) {
+			conn->conn_fd = connect_to_tgt(conn);
+			if (conn->conn_fd >= 0) {
+				conn->conn_state = CS_CONNECT;
+				ev.events = EPOLLOUT;
+				ev.data.ptr = conn;
+				if (epoll_ctl(epollfd, EPOLL_CTL_ADD,
+				    conn->conn_fd, &ev) == -1) {
+					perror("epoll_ctl add");
+					close(conn->conn_fd);
+					conn->conn_fd = -1;
+					rc = -1;
+					break;
+				}
+			}
+		/* we need to close unused connection */
+		} else if (conn->conn_refcount == 0 && conn->conn_fd >= 0) {
+			DBGCONN(conn, "Closing the connection");
+			if (close_conn(conn) != 0) {
+				rc = -1;
+				break;
+			}
+		}
+	}
+	mutex_exit(&conn_list_mtx);
+
+	return (rc);
+}
+
+/*
+ * Try to obtain controller address from zfs property.
+ */
+static int
+get_controller_ip(objset_t *os, char *buf, int len)
+{
+	nvlist_t *props, *propval;
+	char *ip;
+	int error;
+	dsl_pool_t *dp = spa_get_dsl(os->os_spa);
+
+	dsl_pool_config_enter(dp, FTAG);
+	error = dsl_prop_get_all(os, &props);
+	dsl_pool_config_exit(dp, FTAG);
+	if (error != 0)
+		return (error);
+	if (nvlist_lookup_nvlist(props, ZFS_PROP_TARGET_IP, &propval) != 0)
+		return (ENOENT);
+	if (nvlist_lookup_string(propval, ZPROP_VALUE, &ip) != 0)
+		return (EINVAL);
+
+	strncpy(buf, ip, len);
+	nvlist_free(props);
+	return (0);
+}
+
+/*
+ * This gets called whenever a new zinfo is created. We might need to create
+ * a new mgmt connection to iscsi target in response to this event.
+ */
+void
+zinfo_create_cb(zvol_info_t *zinfo, nvlist_t *create_props)
+{
+	char target_host[256];
+	uint16_t target_port;
+	uzfs_mgmt_conn_t *mgmt_conn;
+	zvol_state_t *zv = zinfo->zv;
+	char *delim, *ip;
+	uint64_t val = 1;
+	int rc;
+
+	/* if zvol is being created the zvol property does not exist yet */
+	if (create_props != NULL &&
+	    nvlist_lookup_string(create_props, ZFS_PROP_TARGET_IP, &ip) == 0) {
+		strncpy(target_host, ip, sizeof (target_host));
+	} else {
+		/* get it from zvol properties */
+		if (get_controller_ip(zv->zv_objset, target_host,
+		    sizeof (target_host)) != 0) {
+			/* in case of missing property take the default IP */
+			strncpy(target_host, target_addr, sizeof (target_host));
+			target_port = TARGET_PORT;
+		}
+	}
+
+	delim = strchr(target_host, ':');
+	if (delim == NULL) {
+		target_port = TARGET_PORT;
+	} else {
+		*delim = '\0';
+		target_port = atoi(++delim);
+	}
+
+	mutex_enter(&conn_list_mtx);
+	SLIST_FOREACH(mgmt_conn, &uzfs_mgmt_conns, conn_next) {
+		if (strcmp(mgmt_conn->conn_host, target_host) == 0 &&
+		    mgmt_conn->conn_port == target_port) {
+			/* we already have conn for this target */
+			mgmt_conn->conn_refcount++;
+			zinfo->mgmt_conn = mgmt_conn;
+			mutex_exit(&conn_list_mtx);
+			return;
+		}
+	}
+
+	mgmt_conn = kmem_zalloc(sizeof (*mgmt_conn), KM_SLEEP);
```

Comment:
lets move alloc out of lock

Reply:
will do. makes sense - in 99% of cases it will be needed.